source/particles/generators.cc

```diff
@@ -75,6 +142,263 @@ namespace Particles
+        // the weights of all processes with a lower rank
+        double local_start_weight = 0.0;
+
+        if (const auto tria = dynamic_cast<
+              const parallel::distributed::Triangulation<dim, spacedim> *>(
+              &triangulation))
+          {
+            MPI_Exscan(&local_weight_integral,
+                       &local_start_weight,
+                       1,
+                       MPI_DOUBLE,
+                       MPI_SUM,
+                       tria->get_communicator());
+          }
+
+        // Calculate start id and number of local particles
+        start_particle_id =
+          llround(static_cast<double>(n_particles_to_create) *
+                  local_start_weight / global_weight_integral);
+        n_local_particles =
+          llround(static_cast<double>(n_particles_to_create) *
+                  local_weight_integral / global_weight_integral);
+
+        if (random_cell_selection)
+          {
+            // Uniform distribution on the interval [0,local_weight_integral).
+            // This will be used to randomly select cells for all local
+            // particles.
+            boost::random::uniform_real_distribution<double>
+              uniform_distribution(0.0, local_weight_integral);
+
+            // Loop over all particles to create locally and pick their cells
+            for (types::particle_index current_particle_index = 0;
+                 current_particle_index < n_local_particles;
+                 ++current_particle_index)
+              {
+                // Draw the random number that determines the cell of the
+                // particle
+                const double random_weight =
+                  uniform_distribution(random_number_generator);
+
+                const std::vector<double>::const_iterator selected_cell =
+                  std::lower_bound(accumulated_cell_weights.begin(),
+                                   accumulated_cell_weights.end(),
+                                   random_weight);
+                const unsigned int cell_index =
+                  std::distance(accumulated_cell_weights.begin(),
+                                selected_cell);
+
+                ++particles_per_cell[cell_index];
+              }
+          }
+        else
+          {
+            // Compute number of particles per cell according to the ratio
+            // between their weight and the local weight integral
+            unsigned int          cell_index        = 0;
+            types::particle_index particles_created = 0;
+
+            for (const auto &cell : triangulation.active_cell_iterators())
+              if (cell->is_locally_owned())
+                {
+                  const types::particle_index particles_to_create =
+                    llround(static_cast<double>(n_local_particles) *
+                            accumulated_cell_weights[cell_index] /
+                            local_weight_integral);
```

Comment:
Doesn't this formula yield something like the *cumulative* number of particles to create, from you you get the number of particles to create by taking the difference to the previous cell's value?